### PR TITLE
Restore old cpp frontend

### PR DESCRIPTION
**Pull request by kunjsong01**
Restore old cpp frontend to provide more examples for references.

### CODE REVIEW

**Comment by mikhailramalho** [comment]
Heads up: I think @rafaelsamenezes is also changing the part that enables/disables the old frontends, I think you two should try to converge your efforts here.

**Comment by kunjsong01** [comment]
With the help of the above patches, we are able to collect the logs `esbmc/regression/20220211_233115_esbmc-cpp_cpp`. 

We now get the big picture of the pass/fail rate and the error signatures for cpp old frontend to pass esbmc-cpp/cpp: 

To fully restore cpp olf frontend, we need to fix the following error signatures: 
```
VirtualBox$ egrep "Assertion|###ERROR:" * -rn | rev | cut -d ':' -f 1 | rev | sort | uniq -c
      5  Assertion `decl.get_statement() == "block"' failed.
      2  Assertion `endatptr != atstr.c_str()' failed.
     86  Assertion `type.is_not_nil()' failed.
     33  CONVERSION ERROR
    233  expression statement takes one operand
      7  Message from unrecognized k-induction child process
```

TODOs:
1. (This PR) Continue fixing the errors until we have 100% pass on `esbmc-cpp/cpp`. Start with "**expression statement takes one operand**", which causes 233 failures!
2. (Another PR in the future) Fully bring up `esbmc-cpp/cpp` benchmark with clang-cpp-frontend. Use the old cpp frontend flows to guide the development. 
3. (Another PR in the future) Add C++11 and/or C++14 benchmarks, and fully bring up these benchmarks with clang-cpp-frontend.

**Comment by kunjsong01** [comment]
Update of 04/03/2022: 

54% passing rate in `esbmc-cpp/cpp`, 196 out of 366 tests in this regression suite is passing. 

The stats below shows the error signatures:

```
VirtualBox$ ls 20220304_133136_esbmc-cpp_cpp/* | wc -l
366
VirtualBox$ egrep 'VERIFICATION SUCCESSFUL|VERIFICATION FAILED' 20220304_133136_esbmc-cpp_cpp/ -rn | wc -l
196
VirtualBox$ egrep 'Assertion|###ERROR:' 20220304_133136_esbmc-cpp_cpp/ -rn | rev | cut -d ':' -f 1 | rev | sort | uniq -c
      5  Assertion `decl.get_statement() == "block"' failed.
      6  Assertion `endatptr != atstr.c_str()' failed.
      1  Assertion `is_constant_array2t(elem) && "Sub-member of constant array must be " "constant array"' failed.
     86  Assertion `type.is_not_nil()' failed.
      8  Assertion `type->type_id == trueval->type->type_id' failed.
     33  CONVERSION ERROR
     13 format("Failed to find symbol {} not found", id2string(name)) .c_str()' failed.
      7  Message from unrecognized k-induction child process
```

File: regression/testing_tool.py [hunk 1]
```diff
@@ -34,6 +35,19 @@
...
+# The logs will be kept in the director named as "20220211_073815_<test_name>".
+NOW = datetime.now().strftime("%Y%m%d_%H%M%S")
+LOG_DIR = ""
+LOG_RETENTION = False
+
+# Use old C and C++ frontend
+OLD_FRONTEND = False
+
+# user defined operation mode directory
+OP_MODE_DIR = ""
```
Comment (rafaelsamenezes) [inline regression/testing_tool.py]: Are you sure all those variables need to be global? Can't you refactor into another class? 

Also this could be a PR by itself

File: src/goto-programs/builtin_functions.cpp [hunk 2]
```diff
@@ -515,8 +515,10 @@ void goto_convertt::do_function_call_symbol(
   const irep_idt &identifier = function.identifier();
 
   const symbolt *symbol;
+
   if(ns.lookup(identifier, symbol))
   {
+    ns.lookup(identifier, symbol);
```
Comment (rafaelsamenezes) [inline src/goto-programs/builtin_functions.cpp]: Isn't this duplicated?

File: src/langapi/language_ui.cpp [hunk 3]
```diff
@@ -31,24 +31,14 @@ bool language_uit::parse()
 
 bool language_uit::parse(const std::string &filename)
 {
-  int mode = get_mode_filename(filename);
+  int mode = get_mode_filename(filename, &(config.options));
```
Comment (rafaelsamenezes) [inline src/langapi/language_ui.cpp]: I think that config is a global variable already

File: src/util/message/message.h [hunk 4]
```diff
@@ -30,18 +30,18 @@ class messaget
...
-   * 
+   *
    * @param message string to be printed
    * @param location location where the message happened
    */
   virtual void
   error(const std::string &message, const std::string location = "") const
   {
-    print(VerbosityLevel::Error, message, location);
+    print(VerbosityLevel::Error, "###ERROR: " + message, location);
```
Comment (rafaelsamenezes) [inline src/util/message/message.h]: Thats nice, but it could be in a separate PR

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

File: src/util/message/message.h (hunk 4 above)
Comment (kunjsong01) [reply src/util/message/message.h]: Agreed. I'll split this PR into multiple small PRs.

File: regression/testing_tool.py (hunk 1 above)
Comment (kunjsong01) [reply regression/testing_tool.py]: Agreed. I'll split this PR into multiple small PRs.

File: src/ansi-c/c_main.cpp [hunk 5]
```diff
@@ -297,12 +297,17 @@ bool c_main(
   // Call to main symbol is now in "call"; construct calls to thread library
   // hooks for main thread start and main thread end.
 
+  // For the old frontend, we just use pthread_start_main_hook instead of "c:@F@pthread_start_main_hook" in the old frontend.
+  // The latter is used by clang-c-frontend to verify multi-threaded programs and atomics.
+  // If we want to use the latter, we also need to deal with the symbols referred to from the body of c:@F@pthread_strat_main_hook,
+  // such as __ESBMC_num_threads_running, __ESBMC_num_total_threads, and __ESBMC_atomic_end .etc.
+  // This will become an independent task to add thread support. See clang_c_languaget::final function for example.
```
Comment (mikhailramalho) [inline src/ansi-c/c_main.cpp:304]: Yeah, I think the problem here is that we use clang to gen the names. How do the programs with dynamic memory work with the old frontend? The symbols there will also have different names.

File: src/clang-c-frontend/clang_c_language.cpp [hunk 6]
```diff
@@ -239,10 +239,16 @@ bool clang_c_languaget::typecheck(
...
   contextt new_context(msg);
 
+  // Add symbols in new_context
+  // The symbols here cover AST nodes from internals and source code
   clang_c_convertert converter(new_context, ASTs, msg);
   if(converter.convert())
     return true;
 
+  // Adjust symbol attributes, such as:
+  //  - overwrite macro type with real type
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp:249]: what's a macro type?

File: src/cpp/cpp_language.cpp [hunk 7]
```diff
@@ -145,6 +146,7 @@ bool cpp_languaget::parse(
   std::ostringstream o_preprocessed;
 
   internal_additions(o_preprocessed);
+  //std::cout << o_preprocessed.str() << std::endl;
```
Comment (mikhailramalho) [inline src/cpp/cpp_language.cpp]: remove comments.

File: src/cpp/cpp_language.cpp [hunk 8]
```diff
@@ -201,10 +203,10 @@ bool cpp_languaget::final(contextt &context, const messaget &message_handler)
 
 void cpp_languaget::show_parse(std::ostream &out)
 {
-  for(cpp_parse_treet::itemst::const_iterator it = cpp_parse_tree.items.begin();
-      it != cpp_parse_tree.items.end();
-      it++)
-    show_parse(out, *it);
+  for(const auto &it : cpp_parse_tree.items)
+  {
```
Comment (mikhailramalho) [inline src/cpp/cpp_language.cpp]: no need for curly brackets.

File: src/goto-programs/goto_convert.cpp [hunk 9]
```diff
@@ -419,14 +419,8 @@ void goto_convertt::convert_block(const codet &code, goto_programt &dest)
...
-  if(code.operands().size() != 1)
-  {
-    err_location(code);
-    message_handler.error("expression statement takes one operand\n");
-    abort();
-  }
-
-  exprt expr = code.op0();
+  // Old frontend may have the target operand at op1 rather than op0.
+  exprt expr = (code.operands().size() != 1) ? code.op1() : code.op0();
```
Comment (mikhailramalho) [inline src/goto-programs/goto_convert.cpp:423]: can we change the old frontend to set the target in op0?

I'm not always in favor of fixing the old frontend but I don't think we should change the rest of esbmc because of it, as it might introduce bugs that we simply don't need to fix.

File: src/langapi/language_ui.cpp (hunk 3 above)
Comment (mikhailramalho) [reply src/langapi/language_ui.cpp]: yeah, no need to pass it around.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]

File: src/clang-c-frontend/clang_c_language.cpp (hunk 6 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_language.cpp:249]: A 'macro type' refers to the `symbol.is_macro` as being used in function `clang_c_adjust::adjust_type`: 
```
515     if(symbol.is_macro)
516       type = symbol.type; // overwrite
```
I'm not 100% understanding 'is_macro', hence just briefly explaining the purpose of adjuster in a general manner.
This comment does appear vague and a little misleading. 

Q: What does 'symbol.is_macro' actually do in the code? What's the meaning of it?

File: src/goto-programs/goto_convert.cpp (hunk 9 above)
Comment (kunjsong01) [reply src/goto-programs/goto_convert.cpp:423]: Agreed. This might introduce some unexpected bugs in the future. I'll fix the old frontend instead of goto converter.

Comment (mikhailramalho) [reply src/goto-programs/goto_convert.cpp:423]: yeah, or you can simply ignore it! I mean, the purpose of re-enabling the old frontend is to use the symbol table as a reference, right? If so, you shouldn't worry about the verification being broken later in the verification pipeline.

File: regression/testing_tool.py (hunk 1 above)
Comment (kunjsong01) [reply regression/testing_tool.py]: Fixed and moved to PR#643.

**Comment by kunjsong01** [comment]
Abandoned. 
We are not going to work on the old CPP frontend.

**Comment by fbrausse** [comment]
> We are not going to work on the old CPP frontend.

Does that mean we can remove it?

**Comment by mikhailramalho** [comment]
The only reason the old ones are still around is so that we have a
reference when writing the new ones.

I'm fine with removing them if you all think it's not needed.

It is the last step before killing the old irep/migrate code as well.

**Comment by kunjsong01** [comment]
>Does that mean we can remove it?

We'll remove it when after we bring up the clang-cpp frontend. 
For the time being, the old frontends are just for reference.
